webrtc/videoroom/lib/videoroom/display_engine.ex

```diff
@@ -0,0 +1,183 @@
+  defstruct @enforce_keys
+
+  @doc """
+  Returns new DisplayEngine state that should be passed to subsequent functions.
+
+  `max_display_num` is a maximal number of streams that should be displayed on peer.
+  """
+  @spec new(max_display_num :: non_neg_integer()) :: t()
+  def new(max_display_num) do
+    %__MODULE__{
+      max_display_num: max_display_num,
+      endpoints: %{},
+      display_managers: %{}
+    }
+  end
+
+  @doc """
+  Adds a new endpoint to monitor.
+
+  If other endpoints are going to send video tracks of endpoint `endpoint` then this function should be called.
+  Should be paired with `remove_endpoint/2`.
+  """
+  @spec add_new_endpoint(state :: t(), endpoint :: Endpoint.t()) :: t()
+  def add_new_endpoint(state, %Endpoint{type: :screensharing}), do: state
+
+  def add_new_endpoint(state, endpoint) do
+    display_manager = DisplayManager.new(state.max_display_num)
+    display_manager = setup_display_manager(display_manager, state.endpoints)
+    state = put_in(state.display_managers[endpoint.id], display_manager)
+    put_in(state.endpoints[endpoint.id], endpoint)
+  end
+
+  @doc """
+  Adds a new track that will be sent by other endpoints to the peer.
+  """
+  @spec add_new_track(state :: t(), track_id :: Track.id(), endpoint :: Endpoint.t()) :: t()
+  def add_new_track(state, track_id, endpoint) do
+    track = Endpoint.get_track_by_id(endpoint, track_id)
+    display_managers = add_to_display_managers(track, endpoint, state.display_managers)
+    %{state | display_managers: display_managers}
+  end
+
+  @doc """
+  Updates voice activity status.
+
+  This function should be called each time endpoint with id `endpoint_id` is changing its voice activity state.
+  It returns actions that will enable or disable proper tracks in proper endpoints according to a new status of
+  voice activity for `endpoint_id`.
+  """
+  @spec vad_notification(state :: t(), vad :: boolean(), endpoint_id :: Endpoint.id()) ::
+          {actions :: [], state :: t()}
+  def vad_notification(state, vad, endpoint_id) do
+    {actions, display_managers} =
+      state.display_managers
+      |> Map.delete(endpoint_id)
+      |> Enum.flat_map_reduce(state.display_managers, fn
+        {id, display_manager}, display_managers ->
+          case DisplayManager.update(display_manager, endpoint_id, vad) do
+            {:ok, display_manager} ->
+              {[], Map.put(display_managers, id, display_manager)}
+
+            {{:replace, old_id, new_id}, display_manager} ->
+              old_track_id =
+                Endpoint.get_video_tracks(state.endpoints[old_id])
+                |> List.first()
+                |> case do
+                  %Track{id: id} -> id
+                end
+
+              new_track_id =
+                Endpoint.get_video_tracks(state.endpoints[new_id])
+                |> List.first()
+                |> case do
+                  %Track{id: id} -> id
+                end
```

Comment:
What's the difference between the id returned from DisplayManager and the one taken from Endpoint?

Reply:
DisplayManager returns id of Endpoint which video track should be replaced

Reply:
oh ok, so let's extract 
```elixir
Endpoint.get_video_tracks(state.endpoints[old_id])
|> List.first()
|> case do
  %Track{id: id} -> id
end
```
to a private function

Reply:
Though I'm wondering if DisplayEngine should rather operate on tracks and streams, not only endpoints, so it could return the track IDs directly

Reply:
The first version was based on Track ids but it turned out to be a little easier to operate on Endpoint ids. I would leave it for now and change it while limiting audio to the most loud ones